lib/spack/spack/database.py

```diff
@@ -409,31 +409,83 @@ def _read_suppress_error():
+            # instead, we would perpetuate errors over a reindex.
+
             old_data = self._data
             try:
+                # Initialize data in the reconstructed DB
                 self._data = {}
 
-                # Ask the directory layout to traverse the filesystem.
+                # Start inspecting the installed prefixes
+                processed_specs = set()
+
                 for spec in directory_layout.all_specs():
                     # Try to recover explicit value from old DB, but
-                    # default it to False if DB was corrupt.
-                    explicit = False
+                    # default it to True if DB was corrupt. This is
+                    # just to be conservative in case a command like
+                    # "autoremove" is run by the user after a reindex.
+                    tty.debug(
+                        'RECONSTRUCTING FROM SPEC.YAML: {0}'.format(spec)
+                    )
+                    explicit = True
                     if old_data is not None:
                         old_info = old_data.get(spec.dag_hash())
                         if old_info is not None:
                             explicit = old_info.explicit
 
                     self._add(spec, directory_layout, explicit=explicit)
 
+                    processed_specs.add(spec)
+
+                for key, entry in old_data.items():
+                    # We already took care of this spec using
+                    # `spec.yaml` from its prefix.
+                    if entry.spec in processed_specs:
+                        msg = 'SKIPPING RECONSTRUCTION FROM OLD DB: {0}'
+                        msg += ' [already reconstructed from spec.yaml]'
+                        tty.debug(msg.format(entry.spec))
+                        continue
+
+                    # If we arrived here it very likely means that
+                    # we have external specs that are not dependencies
+                    # of other specs. This may be the case for externally
+                    # installed compilers or externally installed
+                    # applications.
+                    tty.debug(
+                        'RECONSTRUCTING FROM OLD DB: {0}'.format(entry.spec)
+                    )
+                    try:
+                        layout = spack.store.layout
+                        if entry.spec.external:
+                            layout = None
+                        kwargs = {
+                            'spec': entry.spec,
+                            'directory_layout': layout,
+                            'explicit': entry.explicit
+                        }
+                        self._add(**kwargs)
+                        processed_specs.add(entry.spec)
+                    except Exception as e:
+                        # Something went wrong, so the spec was not restored
+                        # from old data
+                        tty.debug(e.message)
```

Comment:
Also for separate PR: it would be good to add something like:

```python
if spack.debug:
    print(traceback.format_exc())
```

So we don't lose the trace.